### PR TITLE
vicadmin redirect/force TLS

### PR DESCRIPTION
**Pull request by akvsh**
Fixes #
Added a redirect for http requests
[specific ci=Group9-VIC-Admin]

### CODE REVIEW

File: cmd/vicadmin/server.go [hunk 1]
```diff
@@ -75,6 +75,82 @@ const (
 	genericErrorMessage    = "Internal Server Error; see /var/log/vic/vicadmin.log for details" // for http errors that shouldn't be displayed in the browser to the user
 )
 
+//wrapper struct around net.Conn for our custom funcs
+type Conn struct {
+	net.Conn
+	b   byte
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: Did we never think of a better name for `f`?

File: cmd/vicadmin/server.go [hunk 2]
```diff
@@ -145,7 +221,8 @@ func (s *server) listen() error {
 		return err
 	}
 
-	s.l = tls.NewListener(innerListener, tlsconfig)
+	s.l = &ReqListener{Listener: innerListener, config: tlsconfig}
+	//s.l = tls.NewListener(innerListener, tlsconfig)
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: delete this commented out line

File: cmd/vicadmin/server.go [hunk 3]
```diff
@@ -158,16 +233,33 @@ func (s *server) AuthenticatedHandle(link string, h http.Handler) {
 	s.Authenticated(link, h.ServeHTTP)
 }
 
+//Redirects HTTP to HTTPS
+func HTTPSRedirectHandle(h http.Handler) http.Handler {
+	redirectToHTTPS := func(w http.ResponseWriter, r *http.Request) {
+		//If TLS is nil, request is not HTTPS, so we must redirect
+		if r.TLS == nil {
+
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: delete this newline :P

File: cmd/vicadmin/server.go [hunk 4]
```diff
@@ -158,16 +233,33 @@ func (s *server) AuthenticatedHandle(link string, h http.Handler) {
...
+			http.Redirect(w, r, fmt.Sprintf("https://%s%s", r.Host, r.RequestURI), http.StatusMovedPermanently)
+			return
+		}
+		h.ServeHTTP(w, r)
+		return
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: This looks like a no-op return that can be deleted

File: cmd/vicadmin/server.go [hunk 5]
```diff
@@ -75,6 +77,80 @@ const (
...
+		}
+		//only one byte read
+		return 1, c.err
+	}
+	//using the default Conn read
+	return c.Conn.Read(b)
+}
+
+//wrapper struct for our custom funcs
+type ReqListener struct {
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: `TLSRedirectListener` ? Maybe something like that?

File: cmd/vicadmin/server.go [hunk 6]
```diff
@@ -75,6 +77,80 @@ const (
...
+			return n + 1, e
+		}
+		//only one byte read
+		return 1, c.err
+	}
+	//using the default Conn read
+	return c.Conn.Read(b)
+}
+
+//wrapper struct for our custom funcs
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: Let's make this comment a bit more descriptive. What does `ReqListener` actually allow us to do?

File: cmd/vicadmin/server.go [hunk 7]
```diff
@@ -75,6 +77,80 @@ const (
...
+type ReqListener struct {
+	net.Listener
+	addr   string
+	config *tls.Config
+}
+
+//override default listener Accept function and add TLS check
+func (l *ReqListener) Accept() (net.Conn, error) {
+	//call default net.listener.Accept first
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: Minor style nit: your comments lack a space after the `//` before the actual comment starts

File: cmd/vicadmin/server.go [hunk 8]
```diff
@@ -75,6 +77,80 @@ const (
...
+	net.Conn
+	b             byte
+	err           error
+	oneTimeSwitch bool
+}
+
+//Returns number of bytes read
+func (c *Conn) Read(b []byte) (int, error) {
+	//oneTimeSwitch bool of first conn starts off as true so we do the TLS check once
+	if c.oneTimeSwitch {
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: How about `c.UncertainTLS` for this flag (sorry to have you keep renaming it)? Then the code has some semantic meaning: "`if c.UncertainTLS` then determine whether TLS is present on this connection or not"

Hopefully you agree that makes the flag's purpose clearer?

File: cmd/vicadmin/server.go [hunk 9]
```diff
@@ -75,6 +77,79 @@ const (
...
+	err          error
+	UncertainTLS bool
+}
+
+// Returns number of bytes read
+func (c *Conn) Read(b []byte) (int, error) {
+	// one time check to determine if TLS is in the connection
+	if c.UncertainTLS {
+		c.UncertainTLS = false
+		b[0] = c.b
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: just for the sake of sanity, before this access we probably want to check the length of `b` and return an error if it's 0

File: cmd/vicadmin/server.go [hunk 10]
```diff
@@ -75,6 +77,82 @@ const (
...
+}
+
+// override default listener Accept function and add TLS check
+func (l *TLSRedirectListener) Accept() (net.Conn, error) {
+	// call default net.listener.Accept first
+	c, err := l.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+	// create slice of size 1 for tru the first byte
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: "for tru the first byte" is this a typo?

File: cmd/vicadmin/server.go [hunk 11]
```diff
@@ -158,16 +235,30 @@ func (s *server) AuthenticatedHandle(link string, h http.Handler) {
...
-	s.mux.Handle(link, gorillacontext.ClearHandler(h))
+	s.mux.Handle(link, gorillacontext.ClearHandler(HTTPSRedirectHandle(h)))
 }
 
 // Enforces authentication on route `link` and runs `handler` on successful auth
 func (s *server) Authenticated(link string, handler func(http.ResponseWriter, *http.Request)) {
 	defer trace.End(trace.Begin(""))
 
 	authHandler := func(w http.ResponseWriter, r *http.Request) {
+
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: remove whitespace

File: cmd/vicadmin/server.go [hunk 12]
```diff
@@ -211,6 +302,7 @@ func (s *server) Authenticated(link string, handler func(http.ResponseWriter, *h
 			}
 
 			// user was authenticated via cert
+
```
Comment (gigawhitlocks) [inline cmd/vicadmin/server.go]: remove whitespace

File: cmd/vicadmin/server.go [hunk 13]
```diff
@@ -75,6 +77,82 @@ const (
 	genericErrorMessage    = "Internal Server Error; see /var/log/vic/vicadmin.log for details" // for http errors that shouldn't be displayed in the browser to the user
 )
 
+// wrapper struct around net.Conn for our custom funcs
```
Comment (jzt) [inline cmd/vicadmin/server.go]: `// Conn is a wrapper struct around net.Conn that implements custom functionality`

https://blog.golang.org/godoc-documenting-go-code

File: cmd/vicadmin/server.go [hunk 14]
```diff
@@ -75,6 +77,82 @@ const (
...
+	return c.Conn.Read(b)
+}
+
+type TLSRedirectListener struct {
+	net.Listener
+	addr   string
+	config *tls.Config
+}
+
+// override default listener Accept function and add TLS check
```
Comment (jzt) [inline cmd/vicadmin/server.go]: `// Accept overrides default listener Accept function and adds a TLS check`

File: cmd/vicadmin/server.go [hunk 15]
```diff
@@ -158,16 +235,30 @@ func (s *server) AuthenticatedHandle(link string, h http.Handler) {
 	s.Authenticated(link, h.ServeHTTP)
 }
 
+// Redirects HTTP to HTTPS
```
Comment (jzt) [inline cmd/vicadmin/server.go]: `// HTTPSRedirectHandle redirects HTTP to HTTPs`

**Pull request merged**